flang/lib/Optimizer/Analysis/AliasAnalysis.cpp

```diff
@@ -522,6 +546,12 @@ AliasAnalysis::Source AliasAnalysis::getSource(mlir::Value v,
         .Case<fir::AllocaOp, fir::AllocMemOp>([&](auto op) {
           // Unique memory allocation.
           type = SourceKind::Allocate;
+          // If there's no DeclareOp, then we need to get the pointer attribute
+          // from the type.  TODO: That case occurs in our test suite
+          // (alias-analysis-2.fir), but does flang currently generate such
+          // code?
+          if (isPointerReference(ty))
+            attributes.set(Attribute::Pointer);
```

Comment:
Can you try remove the 2 changes you made for alias-analysis-2.fir, here and above and set the pointer attribute in
the case for fir::BoxAddrOp:

```
        .Case<fir::BoxAddrOp>([&](auto op) {
          v = op->getOperand(0);
          defOp = v.getDefiningOp();
          if (fir::isPointerType(v.getType()))
            attributes.set(Attribute::Pointer);
          if (mlir::isa<fir::BaseBoxType>(v.getType()))
            followBoxData = true;
        })
```

Reply:
Thanks, check-flang passes with that change.  But can you help me to understand the following points?

1.  In alias-analysis-2.fir, `p1.addr` is `fir.alloca !fir.ptr<f32>`.   I understood that to be pointer non-data (i.e., address of a pointer not of its target).  After applying your suggestion, getSource labels it as non-pointer data instead.  How is the latter correct?
2. As this PR's todo comment asks, when does flang generate something like `fir.alloca !fir.ptr<f32>`?  What semantics does it represent in Fortran?
3. This PR's approach of adding `attributes.set(Attribute::Pointer)` to getSource's AllocaOp case seems consistent with the existing AddrOfOp and dummy arg cases (the two cases that getSource already handles after LoadOp).  Your suggestion makes AllocaOp inconsistent with AddrOfOp and dummy args.  Why is that inconsistency correct?